lib/plugins/EPrints/Plugin/Import/PubRouter.pm

```diff
@@ -802,31 +661,28 @@ sub xml_to_epdata
 	#now process non-primary urls
 	foreach my $doc ( @{$secondary} )
 	{
-		if( $doc->getAttribute( "primary" ) ne "true" )
+		my $document = {};
+		if( $doc->getAttribute( "set_details" ) eq "true" )
 		{
-			my $document = {};
-			my $docFile = $plugin->getDocData( $session, $doc );
-			if( $docFile )
-			{
-				$document->{language} = $code;
-				my %merge = ( %{$docFile}, %{$document} ); #merge doc information with new file
+			$document = $docdata; #copy existing document metadata for this URL		
+		}
+
+		my $docFile = $plugin->getDocData( $session, $doc );
+		if( $docFile )
+		{
+			$document->{language} = $code;
```

Comment:
I note that line 673 does not appear in the preceding similar code block.  
I wonder if a neater approach would be to have 2 stages:
1. Create an ordered array of URLs (primary, secondary)
2. A single loop that does the processing
```
       # download_links need to be processed in order of primary first, so first stage is to order the list
	my @documents = $xml->getElementsByTagNameNS( $namespaces{'pr'}, 'download_link' );
	my $ordered_links = [];
	my $secondary = [];
	foreach my $doc ( @documents )
	{
		if( $doc->getAttribute( "primary" ) eq "true" )
		{
			push @{$ordered_links }, $doc;
		}
		else
		{	
			push @{$secondary}, $doc;
		}
	}
        # Append secondary links to the primary set
        push @{$ordered_links }, @{$secondary};

       # Now process ordered links
	foreach my $doc ( @{$ordered_links } )
	{
		my $document = {};
		if( $doc->getAttribute( "set_details" ) eq "true" )
		{
			$document = $docdata; #copy existing document metadata for this URL		
		}

		my $docFile = $plugin->getDocData( $session, $doc );
		if( $docFile )
		{
			$document->{language} = $code;
			my %merge = ( %{$docFile}, %{$document} ); #merge doc information with new file
	
			if( $merge{mime_type} eq "application/zip" )
			{
				$merge{content} = "other";
			}
			push @{$epdata->{documents}}, \%merge;
		}
	}
```

Reply:
This is once again a much neater way of representing the same code so I've refactored the original - thanks! https://github.com/jisc-services/eprints-plugin-PubRouter-RIOXXplus-Connector/pull/5/commits/5d2140888ed57771e75f7fe23225da064962ed62